### PR TITLE
fix(go): generic composite literal constructor inference (F33)

### LINKED ISSUE #1927: Go: parsing-layer coverage gaps (5 findings)
**Parent epic:** #1919
**Enabled by:** #1921 (heritage helper) — closes F31

5 verified findings. Each box is a discrete fix unit. Legend: ✅ confirmed · 🟡 partial.

- [x] ✅ **F30 [MEDIUM]** Interface method signatures (method_elem) are never captured by any query, so interface methods are not extracted and interface-implementation detection is structurally dead
- [x] ✅ **F31 [MEDIUM]** Struct/interface embedding heritage only matches type_identifier; qualified, generic, and interface-in-interface embeds produce no heritage edge
- [ ] ✅ **F32 [LOW]** Multi-name var/const/field declarations only extract the first name in config-based extractors
- [x] ✅ **F33 [LOW]** Constructor/composite-literal references miss instantiated-generic struct construction (generic_type)
- [x] ✅ **F34 [LOW]** Type-binding param annotation queries omit several common parameter type shapes (channel, array, func, interface, generic)

_Verified against `main` @ `a93ecee0`. Finding IDs use an `F` prefix to avoid colliding with unrelated issue/PR numbers._

### PR DESCRIPTION
**Pull request by evander-wang**
## What changed

Go generic composite literal constructors (e.g. `Box[User]{}`) were not captured by the scope-resolution pipeline. The tree-sitter parser produces `generic_type` nodes for these patterns, but the Go ingestion code only handled plain `type_identifier` nodes in composite literal positions.

This is Go language coverage gap **F33**.

## Why

Generic composite literals are a core Go 1.18+ feature. Without this fix, any codebase using `pkg.T[Type]{}` patterns gets no constructor-inferred type binding, producing missing CALLS edges and incomplete call graphs.

## How

- **Extended tree-sitter queries** (`query.ts`): Added `(generic_type)` to constructor detection patterns in `GO_SCOPE_QUERY`
- **New normalization** (`captures.ts`): `normalizeGenericConstructorCapture()` strips type parameters from `generic_type` nodes to extract the base type name (`Box[User]` → `Box`)
- **Recursive type extraction** (`type-binding.ts`): Enhanced `extractSimpleTypeNameText()` to recursively unwrap `generic_type` and `qualified_type` nesting

## How to verify

```bash
cd gitnexus

# Unit tests — Go type binding (16 tests)
npx vitest run test/unit/scope-resolution/go/go-type-binding.test.ts

# Golden snapshots (9 tests)
npx vitest run test/unit/scope-resolution/go/go-captures-golden.test.ts

# Integration tests — Go resolver (136 tests)
npx vitest run test/integration/resolvers/go.test.ts

# Typecheck
npx tsc --noEmit
```

All pass: 16/16 unit ✅, 9/9 golden ✅, 136/136 integration ✅.

## Risk and rollback

- **Risk:** Low — changes isolated to Go ingestion, no shared infrastructure, backward compatible
- **Rollback:** Revert the commit; non-generic constructor captures are unaffected
- **Follow-ups:** Edge-case tests for nested generics (`Box[Container[User]]{}`) and qualified generics (`pkg.Box[User]{}`) can be added in a separate PR

Refs #1927

### CODE REVIEW

**Review by magyargergo — COMMENTED** [review]
## PR Tri-Review — F33 generic composite-literal constructor inference

**Verdict: production-ready with minor follow-ups.** No correctness, security, or performance regression was found across three review methods; the implementation is correct and the primary cases are well-tested. The remaining items are maintainability/coverage follow-ups, none merge-blocking.

**Methods & engine independence (stated honestly).** GitNexus risk lane + 5 Compound-Engineering personas (correctness, adversarial, maintainability, performance, testing) + **Codex (gpt-5.5, xhigh)**. **7 of the 8 perspectives (6 review lanes + this synthesis) are Claude** — their agreement is "consistent across personas," not independent confirmation. **Only Codex is a different engine**, so Codex-corroborated points carry the real cross-engine weight. The coordinator re-verified every load-bearing claim with real tree-sitter parses in a worktree.

**What's solid (validated, not merely unchallenged):**
- Correctness lane refuted all six probed failure modes with real parses; the adversarial lane could not break it (exotic / multi-arg / `*T` / slice / map type-args, ~2000-level nesting ≈28 ms with no overflow, capture leakage from type-param lists / field types / signatures / `new`/`make`/assertions → **no spurious edges**, container-of-generic correctly non-capturing, no double-emission). Codex independently concluded "no runtime defect."
- **Strictly additive for non-generic Go:** the new `normalizeGenericConstructorCapture` guard cannot fire for `type_identifier` / `qualified_type` nodes, so existing capture output is unchanged — zero regression surface for non-generic repos.
- **Performance:** the new normalization is O(1) per match; the #1848/#1915 hot-path linearity is intact. `bench --check` → scaling ratio **1.144** (budget 1.5); `scaling_budget` correctly left unchanged, only the fingerprint rebaselined.
- **Legacy/registry split is test-accounting, not a production divergence:** Go is registry-primary by default (`MIGRATED_LANGUAGES`), so no production consumer hits the legacy DAG; the `LEGACY_RESOLVER_PARITY_EXPECTED_FAILURES` entry is correctly scoped to the one new test.
- Test runs (correctness/adversarial lanes, worktree sequential parser): Go scope-resolution unit **85/85**, constructor-inference integration **9/9** (incl. the new `Box` test), captures-golden **9/9**; bench fingerprint matches the rebaseline.

**Inline findings (maintainability, non-blocking):**
1. **[P2]** `extractSimpleTypeNameText` is now duplicated verbatim in `captures.ts` and `type-binding.ts` — drift hazard.
2. **[P3]** `extractTypeNode`'s fallback gained `generic_type` but its sibling `extractCompositeLiteralTypeNode` did not — harmless today (the `var x = Box[T]{}` path is rescued by `childForFieldName('type')`; coordinator reproduced → `{name:b, type:Box}`), but a latent inconsistency.

**Codex's independent edge case — investigated, NOT a confirmed defect.** Codex flagged that generic-qualified `models.Box[T]{}` drops the package qualifier at capture time (emits `Box`), while plain `models.Box{}` keeps `models.Box` (coordinator reproduced both). Verified consequence: for the **type-binding / receiver** path this is provably benign — `interpretGoTypeBinding` → `normalizeGoTypeName` strips *both* the qualifier and the `[...]`, so plain and generic both reduce to `Box` (`interpret.ts:37, 93–96`); any shadowing would already affect plain `models.Box{}`, so this PR introduces no regression there. The **reference/CALLS** path is untested under *shadowing* (a package that both declares a local `Box` and instantiates an imported `models.Box[T]`), but Codex itself downgraded this to "coverage, not a runtime defect" after tracing the call path, and the non-shadowing case is proven correct by the new integration test. **Recommend** a shadowed-qualified-generic regression test to close it.

**Coverage gaps (nice-to-have; behaviors verified to work but untested):** `var x = Box[T]{}` (var-spec path), multi-assign generic (`a, b := X{}, Box[int]{}`), nested generic (`Box[Container[int]]{}`), multi-type-arg (`Box[K,V]{}`), qualified generic at the *unit* level, and a negative assertion that the type argument `models.User` does **not** create a spurious `User` edge (adversarial confirmed it doesn't). Separately, the scope-capture **bench synthetic Go source contains no generic literals**, so the F33 path isn't exercised at the 250/800-entity scale — perf for it rests on the O(1)-per-match analysis plus the 1.144 measurement on the small fixture; one generic literal in the bench generator would gate it at scale.

**Merge state & branch hygiene.** `mergeable=MERGEABLE` (no conflicts) but `mergeStateStatus=BLOCKED` → **checks failing**. Branch hygiene: one merge-from-main commit (`3412d4c0`) present but harmless and merge-safe (no overlap with the F33 production files); the other three commits are the F33 work.

**CI.** All substantive gates green — typecheck, lint, format, benchmarks, **scope-resolution parity**, coverage, tree-sitter ABI (all platforms), CodeQL, both Build & Push, macOS platform-sensitive. Failing: **Vercel** (deploy-auth gate, not code) and **`tests / windows-latest (platform-sensitive)` → `CI Gate`**. The Windows failure is a **pre-existing, unrelated flake**: 8 `cli-e2e.test.ts` failures about CLI stdout / fd-1 / EPIPE handling and the `eval-server` `"No indexed repositories found"` startup (`cli-e2e.test.ts:1242/1259/1270/1283/1334/289`) — this PR touches only Go ingestion code, with no CLI/eval-server/stdout surface, so it cannot be the cause.

---
_Automated multi-tool digest — 6 Claude review lanes + Codex (gpt-5.5, xhigh), one independent engine; coordinator-verified against real parses. Treat as input to your judgment, not a substitute — verify before acting._

File: gitnexus/src/core/ingestion/languages/go/captures.ts [hunk 1]
```diff
@@ -340,6 +342,49 @@ function nodeRangeEquals(a: SyntaxNode, b: SyntaxNode): boolean {
...
+      grouped['@reference.name'] = syntheticCapture(
+        '@reference.name',
+        base,
+        extractSimpleTypeNameText(base),
+      );
+    }
+  }
+}
+
+function extractSimpleTypeNameText(node: SyntaxNode): string {
```
Comment (magyargergo) [inline gitnexus/src/core/ingestion/languages/go/captures.ts]: **[P2 · maintainability · code-read]** This `extractSimpleTypeNameText` is byte-identical to the copy in `type-binding.ts:247–257` (same `qualified_type` handling + the new `generic_type` branch). Two copies will drift independently the next time a node type needs unwrapping. `captures.ts` already imports from `type-binding.ts` (line 13), so `export`ing the existing definition there and importing it here removes the duplicate with zero behavior change. Non-blocking.

File: gitnexus/src/core/ingestion/languages/go/type-binding.ts [hunk 2]
```diff
@@ -272,7 +276,9 @@ function extractTypeNode(expr: SyntaxNode): SyntaxNode | null {
   if (expr.type === 'composite_literal') {
     return (
       expr.childForFieldName('type') ??
-      expr.namedChildren.find((c) => ['type_identifier', 'qualified_type'].includes(c.type)) ??
+      expr.namedChildren.find((c) =>
+        ['type_identifier', 'qualified_type', 'generic_type'].includes(c.type),
```
Comment (magyargergo) [inline gitnexus/src/core/ingestion/languages/go/type-binding.ts]: **[P3 · maintainability/consistency · reproduced]** This fallback now includes `generic_type`, but the sibling `extractCompositeLiteralTypeNode` (line 264, the `var`-spec path) still lists only `['type_identifier', 'qualified_type']`. It's harmless today — `var x = Box[T]{}` is rescued by `childForFieldName('type')` (reproduced: yields `{name:'b', type:'Box'}`), so this fallback is effectively dead code — but the asymmetry is a latent trap and the `var`-form generic case currently has no test. Suggest mirroring the list here (or adding a `var`-form generic test). Non-blocking.

File: gitnexus/src/core/ingestion/languages/go/type-binding.ts [hunk 3]
```diff
@@ -261,7 +261,9 @@ function extractCompositeLiteralTypeNode(expr: SyntaxNode): SyntaxNode | null {
...
```
Comment (magyargergo) [inline gitnexus/src/core/ingestion/languages/go/type-binding.ts]: Can we move this array into a set so we won't need to recreate it in every iteration? Plus we can reuse it at line 282.

File: gitnexus/src/core/ingestion/languages/go/interface-impls.ts [hunk 4]
```diff
@@ -324,13 +331,26 @@ function embeddedInterfacesFor(
...
+      const toDelete: string[] = [];
+      for (const id of result) {
+        if (!candidates.has(id)) toDelete.push(id);
+      }
+      for (const id of toDelete) result.delete(id);
+    }
+    if (result.size === 0) return []; // Early exit: no struct has all required methods
   }
-  return best === undefined ? [...indexes.structsById.keys()] : [...best];
+  return result === undefined ? [...indexes.structsById.keys()] : [...result];
```
Comment (magyargergo) [inline gitnexus/src/core/ingestion/languages/go/interface-impls.ts]: If we are carrying a lot of elements in either case, it will result in stackoverflow because of the spread operator.

**Review by magyargergo — CHANGES_REQUESTED** [review]

**Review by magyargergo — APPROVED** [review]
